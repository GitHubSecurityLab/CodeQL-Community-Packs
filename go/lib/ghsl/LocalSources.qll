private import go

module LocalSources {
  private import semmle.go.dataflow.DataFlow
  private import semmle.go.dataflow.TaintTracking
  private import semmle.go.Scopes
  
  abstract class Range extends DataFlow::Node { }

// ========== Sources ==========

abstract class Sources extends DataFlow::Node { }

// ----------------------------------------------------
// Used for finding Selections or Calls for Go imports 
// ----------------------------------------------------

//class UseOfGoImports extends Sources { 
  //UseOfGoImports () {
    //exists ( ValueEntity read, 
             //DataFlow::Package pkg | 
             //read.getScope().getEntity(_) = pkg.getScope().getEntity(_)
             //and ( this.toString().regexpMatch("selection of.*")
              //or this.toString().regexpMatch("call to .*") )
    //)
  //}
//}

// ----------------------------------------------------

class OsCmd extends LocalSources::Range {
  OsCmd() {
    exists ( ValueEntity read, 
      DataFlow::Package pkg | 
      read.getScope().getEntity(_) = pkg.getScope().getEntity(_)
      and this.toString() = "selection of Run"
    ) 
  }
}

class OsExec extends LocalSources::Range {
  OsExec() {
    exists ( ValueEntity read, 
             DataFlow::Package pkg | 
             read.getScope().getEntity(_) = pkg.getScope().getEntity(_)
             and this.toString() = "selection of Command"
    )
  }
}

class OsArgs extends LocalSources::Range { 
  OsArgs() {
    exists ( ValueEntity read, 
             DataFlow::Package pkg | 
             read.getScope().getEntity(_) = pkg.getScope().getEntity(_)
             and this.toString() = "selection of Args"
    )
  }
}

// Not currently working (need a test case)
//class OsGetenv extends Sources, DataFlow::CallNode {
  //OsGetenv() {
    //// https://pkg.go.dev/os#Getenv
    //this.getTarget().hasQualifiedName(package("os", ""), "Getenv")
    //or
    //// https://pkg.go.dev/os#Environ
    //this.getTarget().hasQualifiedName(package("os", ""), "Environ")
  //}
//}

  // https://pkg.go.dev/flag
class Flag extends LocalSources::Range {
    Flag() {
      exists ( ValueEntity read, 
               DataFlow::Package pkg | 
               read.getScope().getEntity(_) = pkg.getScope().getEntity(_)
               and 
               ( this.toString() = "selection of String"
               or this.toString() = "selection of Parse" )
      )
    }
}
}