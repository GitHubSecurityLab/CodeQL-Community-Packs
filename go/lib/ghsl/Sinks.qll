private import go
private import semmle.go.dataflow.DataFlow
private import semmle.go.security.CommandInjectionCustomizations
private import semmle.go.security.OpenUrlRedirectCustomizations
private import semmle.go.security.ReflectedXssCustomizations
private import semmle.go.security.RequestForgeryCustomizations
private import semmle.go.security.SqlInjectionCustomizations
private import semmle.go.security.UnsafeUnzipSymlinkCustomizations
private import semmle.go.security.XPathInjectionCustomizations
private import semmle.go.security.ZipSlipCustomizations

/**
 * List of all the sinks that we want to check.
 */
class AllSinks extends DataFlow::Node {
  private string sink;

  AllSinks() {
    this instanceof CommandInjection::Sink and
    sink = "command-injection"
    or
    this instanceof OpenUrlRedirect::Sink and
    sink = "open-url-redirect"
    or
    this instanceof ReflectedXss::Sink and
    sink = "reflected-xss"
    or
    this instanceof RequestForgery::Sink and
    sink = "request-forgery"
    or
    this instanceof SqlInjection::Sink and
    sink = "sql-injection"
    or
    this instanceof UnsafeUnzipSymlink::EvalSymlinksSink and
    sink = "unsafe-unzip"
    or
    this instanceof XPathInjection::Sink and
    sink = "xpath-injection"
    or
    this instanceof ZipSlip::Sink and
    sink = "zip-slip"
  }

  /**
   * Gets the sink sink type.
   */
  string sinkType() { result = sink }
}
