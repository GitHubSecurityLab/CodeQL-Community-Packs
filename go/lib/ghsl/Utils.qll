private import go
private import semmle.go.dataflow.DataFlow
private import semmle.go.dataflow.TaintTracking
private import semmle.go.frameworks.stdlib.Fmt

/**
 * Find Node at Location
 */
predicate filterByLocation(DataFlow::Node node, string relative_path, int linenumber) {
  node.getLocation().getFile().getRelativePath() = relative_path and
  node.getLocation().getStartLine() = linenumber
}

/**
 * List of all the souces
 */
class AllSources extends DataFlow::Node {
  private string threatmodel;

  AllSources() {
    this instanceof RemoteFlowSource::Range and
    threatmodel = "remote"
    or
    this instanceof LocalSources and
    threatmodel = "local"
  }

  /**
   * Gets the source threat model.
   */
  string getThreatModel() { result = threatmodel }
}

/**
 * Local sources
 */
class LocalSources extends DataFlow::Node {
  LocalSources() {
    this.(SourceNode).getThreatModel() = "local"
  }
}

class DynamicStrings extends DataFlow::Node {
    DynamicStrings() {
        // fmt format string
        exists(Fmt::Sprinter formatter |
            this = formatter.getACall()
        )
        or
        // binary expression
        exists(BinaryExpr expr |
            this.asExpr() = expr.getLeftOperand() and
            expr.getOperator() = "+"
        )
    }
}