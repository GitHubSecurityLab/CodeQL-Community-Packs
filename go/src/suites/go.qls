- description: "GitHub's Community Packs Extended GoLang Suite"

- queries: '.'
  from: githubsecuritylab/codeql-go-queries

- include:
    kind:
    - problem
    - path-problem
    precision:
    - very-high
    - high
    - medium
    - low

# Remove debugging, and audit queries
- exclude:
    tags contain:
      - debugging
      - audit

# Remove local testing folders
- exclude:
    query path:
      - /testing\/.*/
