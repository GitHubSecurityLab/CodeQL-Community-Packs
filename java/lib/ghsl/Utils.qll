/**
 * A collection of utility predicates and classes for the Java library.
 */

private import semmle.code.java.dataflow.DataFlow
private import semmle.code.java.dataflow.ExternalFlow
private import semmle.code.java.dataflow.FlowSources
// Sinks
private import semmle.code.java.security.QueryInjection
private import semmle.code.java.security.CommandLineQuery
private import semmle.code.java.security.LdapInjection
private import semmle.code.java.security.LogInjection
private import semmle.code.java.security.OgnlInjection
private import semmle.code.java.security.RequestForgery
private import semmle.code.java.security.TemplateInjection

/**
 * Filter nodes by its location (relative path or base name).
 */
bindingset[relative_path]
predicate findByLocation(DataFlow::Node node, string relative_path, int linenumber) {
  node.getLocation().getFile().getRelativePath().matches(relative_path) and
  node.getLocation().getStartLine() = linenumber
}

/**
 * This will only show sinks that are callable (method calls)
 */
predicate isCallable(DataFlow::Node sink) { sink.asExpr() instanceof MethodCall }

/**
 * Check if the source node is a method parameter.
 */
predicate checkSource(DataFlow::Node source) {
  exists(source.asParameter())
  or
  source.asExpr() instanceof MethodCall
}

/**
 * Local sources
 */
class LocalSources = LocalUserInput;

/**
 * List of all the souces
 */
class AllSources extends DataFlow::Node {
    private string threadmodel;

  AllSources() {
    this instanceof LocalUserInput and
    threadmodel = "local"
    or
    this instanceof RemoteFlowSource and
    threadmodel = "remote"
    or
    this instanceof ActiveThreatModelSource
    and
    threadmodel = this.(SourceNode).getThreatModel()
  }

  /**
   * Gets the source threat model.
   */
  string getThreatModel() {
    result = threadmodel
  }
}

/**
 * List of all the sinks that we want to check.
 */
class AllSinks extends DataFlow::Node {
  private string sink;

  AllSinks() {
    this instanceof QueryInjectionSink
    and 
    sink = "QueryInjectionSink"
    or
    this instanceof CommandInjectionSink
    and
    sink = "CommandInjectionSink"
    or
    this instanceof LdapInjectionSink
    and
    sink = "LdapInjectionSink"
    or
    this instanceof LogInjectionSink
    and
    sink = "LogInjectionSink"
    or
    this instanceof OgnlInjectionSink
    and
    sink = "OgnlInjectionSink"
    or
    this instanceof RequestForgerySink
    and
    sink = "RequestForgerySink"
    or
    this instanceof TemplateInjectionSink
    and
    sink = "TemplateInjectionSink"
    or
    // All MaD sinks
    sinkNode(this, _)
    and
    sink = "MaD"
  }
  
  /**
   * Gets the sink sink type.
   */
  string sinkType() {
    result = sink
  }
}
