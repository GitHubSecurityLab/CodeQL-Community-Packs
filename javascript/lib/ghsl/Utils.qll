/**
 * A collection of utility predicates and classes for JavaScript
 */
private import javascript
private import semmle.javascript.security.dataflow.CommandInjectionCustomizations
private import semmle.javascript.security.dataflow.CodeInjectionCustomizations
private import semmle.javascript.security.dataflow.LogInjectionQuery as LogInjection
private import semmle.javascript.security.dataflow.NosqlInjectionCustomizations
private import semmle.javascript.security.dataflow.Xss as Xss
private import semmle.javascript.security.dataflow.XxeCustomizations


/**
 * Filter results to a specific file and line number
 * 
 *  **Examples:**
 * 
 *  ```
 *  filterByLocation(sources, "db.js", 1)
 *  // or we don't care about the line numbers
 *  filterByLocation(sources, "db.js", _)
 *  ```
 */
predicate filterByLocation(DataFlow::Node node, string relative_path, int linenumber) {
  node.getLocation().getFile().getRelativePath() = relative_path and
  node.getLocation().getStartLine() = linenumber
}


/**
 * All Sources (Remote and Local)
 */
class AllSources extends DataFlow::Node  {
    private string threadmodel;

    AllSources() {
        this instanceof RemoteSources and
        threadmodel = "remote" or
        this instanceof LocalSources and
        threadmodel = "local"
    }

  /**
   * Gets the source threat model.
   */
  string getThreatModel() {
    result = threadmodel
  }
}

/**
 * Remote Sources (HTTP frameworks, etc)
 */
class RemoteSources extends ThreatModelSource {
  RemoteSources() { this.getThreatModel() = "remote" }
}

/**
 * Local Sources (CLI arguments, Filesystem, etc)
 */
class LocalSources extends ThreatModelSource {
  LocalSources() { this.getThreatModel() = "local" }
}

/**
 * List of all sinks
 */
class AllSinks extends DataFlow::Node {
  private string sink;

  AllSinks() {
    this instanceof CodeInjection::Sink and
    sink = "code-injection" or
    this instanceof CommandInjection::Sink and
    sink = "command-injection" or
    this instanceof LogInjection::Sink and
    sink = "log-injection" or
    this instanceof NosqlInjection::Sink and
    sink = "nosql-injection" or
    this instanceof Xss::Shared::Sink and
    sink = "xss" or
    this instanceof Xxe::Sink and
    sink = "xxe"
  }

  /**
   * Gets the sink threat model.
   */
  string sinkType() {
    result = sink
  }
}