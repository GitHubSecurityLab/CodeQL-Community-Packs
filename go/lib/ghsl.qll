import go
import ghsl.Utils
import ghsl.LocalSources
import ghsl.Sinks