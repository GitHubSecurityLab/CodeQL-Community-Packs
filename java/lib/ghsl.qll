import ghsl.LocalSources
// Export utils
import ghsl.Utils