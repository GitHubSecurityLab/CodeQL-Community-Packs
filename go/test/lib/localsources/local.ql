import go
import ghsl.Utils
import ghsl.LocalSources

query predicate remoteSources(DataFlow::ExprNode node) { node instanceof RemoteFlowSource::Range }

query predicate localSources(DataFlow::ExprNode node) { node instanceof LocalSources::Range }
