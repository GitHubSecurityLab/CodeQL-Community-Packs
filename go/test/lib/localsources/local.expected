| cmd/flag.go:10:13:10:23 | selection of String |
| cmd/flag.go:10:13:10:50 | call to String |
| cmd/flag.go:18:2:18:11 | selection of Parse |
| cmd/flag.go:24:23:24:31 | selection of Args |
| cmd/go_os.go:9:10:9:16 | selection of Args |
| cmd/go_os.go:13:9:13:20 | call to Environ |
| cmd/go_os.go:17:11:17:27 | call to Getenv |
